src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,392 @@
+position within the subwindows.
+
+This class calculates a statistical profile S over all $S_i$, i.e. the value of
+S at position i is calculated by performing a certian calculation, e.g. an
+average, over all values of $S_i$ at position i.
+"""
+
+from __future__ import annotations
+
+import logging
+import math
+from datetime import datetime, timedelta
+from typing import List, Tuple
+
+import numpy as np
+from numpy.typing import NDArray
+
+from frequenz.sdk.timeseries._moving_window import MovingWindow
+
+log = logging.getLogger(__name__)
+
+
+class PeriodicFeatureExtractor:
+    """
+    A tool that extracts a timeseries feature out of historical data.
+
+    This class is creating a profile from periodically occurring windows in a
+    buffer of historical data.
+
+    The profile is created out of all windows that are fully contained in the
+    underlying buffer with the same start and end time modulo a fixed period.
+
+    Note:
+        The oldest window or the window that is currently overwritten in the
+        MovingWindow is not considered in the profile.
+
+    Note:
+        When constructing a PeriodicFeatureExtractor object the `MovingWindow`
+        size has to be a integer multiple of the period.
+
+    Example:
+        ```python
+        from datetime import datetime, timedelta, timezone
+
+        moving_window = MovingWindow(
+            size=timedelta(days=35),
+            resampled_data_recv=microgrid.logical_meter().grid_power.new_receiver(),
+            input_sampling_period=timedelta(seconds=1),
+        )
+
+        feature_extractor = PeriodicFeatureExtractor(
+            moving_window = moving_window,
+            period=timedelta(days=7),
+        )
+
+        now = datetime.now(tzinfo=timezone.utc)
+
+        # create a daily weighted average for the next 24h
+        avg_24h = feature_extractor.avg(
+            now,
+            now + timedelta(hours=24),
+            weights=[0.1, 0.2, 0.3, 0.4]
+        )
+
+        # create a daily average for Thursday March 23 2023
+        th_avg_24h = feature_extractor.avg(datetime(2023,3,23), datetime(2023,3,24))
+        ```
+    """
+
+    def __init__(
+        self,
+        moving_window: MovingWindow,
+        period: timedelta,
+    ) -> None:
+        """
+        Initialize a PeriodicFeatureExtractor object.
+
+        Args:
+            moving_window: The MovingWindow that is used for the average calculation.
+            period: The distance between two succeeding intervals.
+
+        Raises:
+            ValueError: If the MovingWindow size is not a integer multiple of the period.
+        """
+        self._moving_window = moving_window
+        self._buffer = moving_window._buffer
+
+        self._sampling_period = self._moving_window.sampling_period.total_seconds()
+        """The sampling_period as integer to use it for indexing of samples."""
+
+        self._period = int(period.total_seconds() / self._sampling_period)
```

Comment:
Does that work correctly if period is not a multiple of sampling period?

Reply:
oh, I see you check this below. For unexpected behavior I would raise an exception or do you see valid use-cases for this?

Reply:
I didn't want to add unnecessary restrictions and I haven't thought it fully through. I suggest we address this by adding tests in a follow up PR.